modules/shortcodes/youtube.php

```diff
@@ -365,16 +294,203 @@ function youtube_id( $url ) {
+			esc_url( $url )
+		);
+	}
+
+	$sanitized_url          = youtube_sanitize_url( $url );
+	$parsed_url             = wp_parse_url( $sanitized_url );
+	$query_args             = jetpack_shortcode_youtube_query_args( $parsed_url );
+	list( $width, $height ) = jetpack_shortcode_youtube_dimensions( $query_args );
+	return sprintf(
+		'<amp-youtube data-videoid="%s" layout="responsive" width="%d" height="%d"></amp-youtube>',
+		esc_attr( $video_id ),
+		absint( $width ),
+		absint( $height )
+	);
+}
+
+/**
+ * Gets the dimensions of the [youtube] shortcode.
+ *
+ * Calculates the width and height, taking $content_width into consideration.
+ *
+ * @since 8.0.0
+ *
+ * @param array $query_args The query args of the URL.
+ *
+ * @return array The width and height of the shortcode.
+ */
+function jetpack_shortcode_youtube_dimensions( $query_args ) {
+	global $content_width;
+
+	$input_w = ( isset( $query_args['w'] ) && intval( $query_args['w'] ) ) ? intval( $query_args['w'] ) : 0;
+	$input_h = ( isset( $query_args['h'] ) && intval( $query_args['h'] ) ) ? intval( $query_args['h'] ) : 0;
+
+	// If we have $content_width, use it.
+	if ( ! empty( $content_width ) ) {
+		$default_width = $content_width;
+	} else {
+		// Otherwise get default width from the old, now deprecated embed_size_w option.
+		$default_width = get_option( 'embed_size_w' );
+	}
+
+	// If we don't know those 2 values use a hardcoded width.
+	if ( empty( $default_width ) ) {
+		$default_width = 640;
+	}
+
+	if ( $input_w > 0 && $input_h > 0 ) {
+		$w = $input_w;
+		$h = $input_h;
+	} elseif ( 0 === $input_w && 0 === $input_h ) {
+		if ( isset( $query_args['fmt'] ) && intval( $query_args['fmt'] ) ) {
+			$w = ( ! empty( $content_width ) ? min( $content_width, 480 ) : 480 );
+		} else {
+			$w = ( ! empty( $content_width ) ? min( $content_width, $default_width ) : $default_width );
+			$h = ceil( ( $w / 16 ) * 9 );
+		}
+	} elseif ( $input_w > 0 ) {
+		$w = $input_w;
+		$h = ceil( ( $w / 16 ) * 9 );
+	} else {
+		if ( isset( $query_args['fmt'] ) && intval( $query_args['fmt'] ) ) {
+			$w = ( ! empty( $content_width ) ? min( $content_width, 480 ) : 480 );
+		} else {
+			$w = ( ! empty( $content_width ) ? min( $content_width, $default_width ) : $default_width );
+			$h = $input_h;
+		}
+	}
+
+	/**
+	 * Filter the YouTube player width.
+	 *
+	 * @module shortcodes
+	 *
+	 * @since 1.1.0
+	 *
+	 * @param int $w Width of the YouTube player in pixels.
+	 */
+	$w = (int) apply_filters( 'youtube_width', $w );
+
+	/**
+	 * Filter the YouTube player height.
+	 *
+	 * @module shortcodes
+	 *
+	 * @since 1.1.0
+	 *
+	 * @param int $h Height of the YouTube player in pixels.
+	 */
+	$h = (int) apply_filters( 'youtube_height', $h );
+
+	return array( $w, $h );
+}
+
+/**
+ * Gets the video ID from the URL.
+ *
+ * @since 8.0.0
+ *
+ * @param string $url The YouTube URL.
+ *
+ * @return string|bool The video ID based on that URL, or false.
+ */
+function get_amp_youtube_id_from_url( $url ) {
```

Comment:
We currently have a `jetpack_get_youtube_id` function one can use to get an ID from a YouTube URL. Do you think it would be helpful to you here, or are there changes from `get_amp_youtube_id_from_url` you think we should bring to `jetpack_get_youtube_id`? In any case, I think it'd be nice if we only had one function to take care of that.

https://github.com/Automattic/jetpack/blob/792b26b5539d07cc35fdd93567942f2ad481eef2/functions.compat.php#L20

Reply:
Great point, I didn't see that function.

Reply:
https://github.com/Automattic/jetpack/pull/14010/commits/d7c05647435878c32ee1bece9814c4c933bad32e uses `jetpack_get_youtube_id()` instead of the custom function ported from AMP. 

There doesn't look to be an regression from this. It even passed this unit test when substituting `jetpack_get_youtube_id` for `get_amp_youtube_id_from_url()` on [this line](https://github.com/Automattic/jetpack/pull/14010/commits/d7c05647435878c32ee1bece9814c4c933bad32e#diff-7835daad42d569641fcb9ae94cce9f51L280).